### PR TITLE
Fps Counter

### PR DESCRIPTION
**Pull request by jojolepro**
Adds a FPSCounter struct and a system to amethyst. Usage in rust documentation.

### CODE REVIEW

**Comment by Rhuagh** [comment]
Please run cargo fmt.

**Comment by jojolepro** [comment]
I did already, look at the commit names.
I did
rustup update nightly
rustup default nightly
cargo fmt
rutup default stable
cargo build

File: src/util/datastructure.rs [hunk 1]
```diff
@@ -0,0 +1,28 @@
+use std::collections::VecDeque;
```
Comment (Rhuagh) [inline src/util/datastructure.rs]: Could we rename this file to something that says what it contains? Like circular_buffer.rs or such.

File: src/util/fpscounter.rs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        for e in self.buf.queue().iter(){
+            total += *e;
+        }
+        1.0 / (total / self.buf.queue().len() as f32)
+    }
+}
+
+///FPSCounterSystem
+///Add this system to your game to calculate FPS
+pub struct FPSCounterSystem;
```
Comment (Rhuagh) [inline src/util/fpscounter.rs]: Can we keep ECS code under the ecs package ?
Put this under ecs::util or ecs::fps or something like that.

**Comment by jojolepro** [comment]
I'm not too sure if I made the mod.rs files properly.

File: src/util/fpscounter.rs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+use std::collections::VecDeque;
+use amethyst::ecs::System;
+
+///The FPSCounter resource needed by the FPSCounterSystem
+///Add it to your resources to be able to use the FPSCounterSystem
+///world.add_resource(FPSCounter::new(20));
+///Usage example:
+///Get the FPSCounter resource.
+///println!("Cur FPS: {}, Sampled: {}",counter.frame_fps(),counter.sampled_fps());
+pub struct FPSCounter{
```
Comment (Xaeroxe) [inline src/util/fpscounter.rs]: Move this into src/ecs/util/resources in order to conform with the rest of the crate structure.

File: src/util/datastructure.rs (hunk 1 above)
Comment (jojolepro) [reply src/util/datastructure.rs]: Fixed

File: src/util/fpscounter.rs (hunk 2 above)
Comment (jojolepro) [reply src/util/fpscounter.rs]: Fixed (I think)

File: src/util/circular_buffer.rs [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    pub fn new<B>(size:usize)->CircularBuffer<B>{
+        CircularBuffer{
+            queue: VecDeque::with_capacity(size),
+            size:size,
+        }
+    }
+    //front to back
+    ///Add a value to the CircularBuffer
+    pub fn push(&mut self,elem:A){
+        if self.queue.len() < self.size {
```
Comment (msiglreith) [inline src/util/circular_buffer.rs]: This logic can be simplified by moving the `push_back` out of the if/else construct

File: src/ecs/util/systems.rs [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+     );
+    fn run(
+        &mut self,
+        (
+         time,
+         mut counter,
+         ):
+         Self::SystemData,
+    ){
+        counter.push(duration_to_secs(time.delta_time));
```
Comment (msiglreith) [inline src/ecs/util/systems.rs]: Directly storing the `Duration` would be nicer imo.

Comment (jojolepro) [reply src/ecs/util/systems.rs]: As parameter of counter.push or as a variable in the system?

Comment (jojolepro) [reply src/ecs/util/systems.rs]: @msiglreith

File: src/util/circular_buffer.rs (hunk 4 above)
Comment (jojolepro) [reply src/util/circular_buffer.rs]: I have no idea why I did such a stupid error. Thanks for pointing it out!

File: src/ecs/util/systems.rs (hunk 5 above)
Comment (msiglreith) [reply src/ecs/util/systems.rs]: Sry for the delayed response. Storing `Duration`s inside the `FpsCounter` instead of `f32`.

Comment (jojolepro) [reply src/ecs/util/systems.rs]: Lost of memory, lost of performance.
I'd have to put the durations back into f32 each time I want the average fps each tick.

Comment (jojolepro) [reply src/ecs/util/systems.rs]: Also, I doubt anyone will ever interact directly with it without using the System

**Comment by jojolepro** [comment]
@Rhuagh @msiglreith @omni-viral @LucioFranco For review! Thanks!

File: src/ecs/util/systems.rs (hunk 5 above)
Comment (stenverbois) [reply src/ecs/util/systems.rs]: You would be calculating the average of a u64 and a u32 number, and converting to f32 once in `sampled_fps`. I don't see how that is less performant, especially when the user doesn't request the sampled fps every frame.

Comment (omni-viral) [reply src/ecs/util/systems.rs]: Actually you can store single u64 that represents nanoseconds. u64 can store more than hundred years of nanoseconds.
30 years is about 1'000'000'000'000'000'000 nanoseconds

**Comment by jojolepro** [comment]
My push method is now bugged because I cleaned in the wrong order :/

**Comment by anderejd** [comment]
@jojolepro You dont need to switch between stable and nightly, you can just do `cargo +nightly fmt`

**Comment by jojolepro** [comment]
Thanks @anderejd

File: src/util/time.rs [hunk 6]
```diff
@@ -0,0 +1,10 @@
+//! Util time conversion functions
+use std::time::Duration;
+///Converts a Duration to the time in seconds.
+pub fn duration_to_secs(duration: &Duration) -> f32 {
+    duration.as_secs() as f32 + (duration.subsec_nanos() as f32 / 1.0e9)
+}
+///Converts a Duration to nanoseconds
+pub fn duration_to_nanos(duration: &Duration) -> u64 {
+    (duration.as_secs() as f64 * 1.0e9) as u64 + duration.subsec_nanos() as u64
```
Comment (omni-viral) [inline src/util/time.rs]: Too many conversions

Comment (stenverbois) [reply src/util/time.rs]: Indeed, no need for floating point here. Just use ```duration.as_secs() * 1_000_000_000 + ...```

**Comment by Xaeroxe** [comment]
The only removals in this PR are formatting changes, and there's a lot of commits here so I plan on squashing this as soon as either appveyor or travis succeeds, as this isn't platform specific code.

File: src/ecs/util/resources.rs [hunk 7]
```diff
@@ -0,0 +1,40 @@
+//! Util Resources
+use util::circular_buffer::CircularBuffer;
+
+///The FPSCounter resource needed by the FPSCounterSystem
+///Add it to your resources to be able to use the FPSCounterSystem
+///world.add_resource(FPSCounter::new(20));
+///Usage example:
+///Get the FPSCounter resource.
+///println!("Cur FPS: {}, Sampled: {}",counter.frame_fps(),counter.sampled_fps());
```
Comment (torkleyy) [inline src/ecs/util/resources.rs]: This should really be in a codeblock

Comment (jojolepro) [reply src/ecs/util/resources.rs]: Fixed,

**Comment by Xaeroxe** [comment]
@torkleyy Looks like this was fixed by @jojolepro , can you r+ this if you like it?

File: src/ecs/util/resources.rs [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+/// The FPSCounter resource needed by the FPSCounterSystem
+/// # Examples
+/// Add it to your resources to be able to use the FPSCounterSystem
+/// ~~~no_run
+/// world.add_resource(FPSCounter::new(20));
+/// ~~~
+/// Usage example:
+/// ~~~no_run
+/// Get the FPSCounter resource from the System or from the world.
```
Comment (torkleyy) [inline src/ecs/util/resources.rs]: This won't compile; please use a comment and create a counter; also don't forget to import the items you're using.

File: src/ecs/util/systems.rs [hunk 9]
```diff
@@ -0,0 +1,20 @@
+//! Util Systems
+use ecs::util::resources::FPSCounter;
+use specs::{Fetch, FetchMut, System};
+use timing::Time;
+use util::time::duration_to_nanos;
+///FPSCounterSystem
+///Add this system to your game to calculate FPS
+/// # Examples
```
Comment (torkleyy) [inline src/ecs/util/systems.rs]: Should be `## Examples`

File: src/ecs/util/systems.rs [hunk 10]
```diff
@@ -0,0 +1,20 @@
+//! Util Systems
+use ecs::util::resources::FPSCounter;
+use specs::{Fetch, FetchMut, System};
+use timing::Time;
+use util::time::duration_to_nanos;
+///FPSCounterSystem
```
Comment (torkleyy) [inline src/ecs/util/systems.rs]: Please add a newline before the doc coment and leave a space at the beginning of each line in comments.

File: src/util/circular_buffer.rs [hunk 11]
```diff
@@ -0,0 +1,29 @@
+//! CircularBuffer
+use std::collections::VecDeque;
+///A CircularBuffer that drops the oldest element inserted when full.
```
Comment (torkleyy) [inline src/util/circular_buffer.rs]: Please add one newline after the mod comment and one after the import.

File: src/util/time.rs [hunk 12]
```diff
@@ -0,0 +1,10 @@
+//! Util time conversion functions
+use std::time::Duration;
+///Converts a Duration to the time in seconds.
+pub fn duration_to_secs(duration: &Duration) -> f32 {
+    duration.as_secs() as f32 + (duration.subsec_nanos() as f32 / 1.0e9)
+}
+///Converts a Duration to nanoseconds
```
Comment (torkleyy) [inline src/util/time.rs]: Again, let's add a newline here.

File: src/util/time.rs [hunk 13]
```diff
@@ -0,0 +1,10 @@
+//! Util time conversion functions
+use std::time::Duration;
+///Converts a Duration to the time in seconds.
+pub fn duration_to_secs(duration: &Duration) -> f32 {
```
Comment (torkleyy) [inline src/util/time.rs]: Stylistically, a copy would be preferable if the type is `Copy`.

File: src/util/circular_buffer.rs [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+pub struct CircularBuffer<A> {
+    queue: VecDeque<A>,
+}
+impl<A> CircularBuffer<A> {
+    ///Creates a new CircularBuffer with fixed size
+    pub fn new<B>(size: usize) -> CircularBuffer<B> {
+        CircularBuffer {
+            queue: VecDeque::with_capacity(size),
+        }
+    }
```
Comment (torkleyy) [inline src/util/circular_buffer.rs]: Please add a newline here.

File: src/ecs/util/systems.rs (hunk 9 above)
Comment (jojolepro) [reply src/ecs/util/systems.rs]: src/app.rs
    /// # Examples

**Comment by jojolepro** [comment]
Will fix within the next hour or so, thanks for reviewing!

File: src/util/circular_buffer.rs (hunk 14 above)
Comment (jojolepro) [reply src/util/circular_buffer.rs]: Where?

File: src/util/circular_buffer.rs (hunk 11 above)
Comment (jojolepro) [reply src/util/circular_buffer.rs]: Fixed

File: src/ecs/util/systems.rs (hunk 9 above)
Comment (jojolepro) [reply src/ecs/util/systems.rs]: Fixed

File: src/ecs/util/systems.rs (hunk 10 above)
Comment (jojolepro) [reply src/ecs/util/systems.rs]: Fixed

File: src/util/time.rs (hunk 12 above)
Comment (jojolepro) [reply src/util/time.rs]: Fixed

File: src/util/time.rs (hunk 13 above)
Comment (jojolepro) [reply src/util/time.rs]: Fixed, thanks for the tip, I wasn't aware that Duration was Copy

File: src/ecs/util/resources.rs (hunk 8 above)
Comment (jojolepro) [reply src/ecs/util/resources.rs]: Fixed, its just a comment and not and actual example, as I can't get a world instance without creating a amethyst context.

**Comment by jojolepro** [comment]
Fixed everything you asked for @torkleyy

**Comment by Xaeroxe** [comment]
bors works in a queue and the travis fix has been queued, so it's safe to queue this too.

bors r+

**Comment by Xaeroxe** [comment]
bors r-

I'll fix the conflicts I just made.

**Comment by Xaeroxe** [comment]
Overriding.

**Pull request merged**